crates/telio-dns/src/nameserver.rs

```diff
@@ -414,23 +466,73 @@ impl RequestInfo {
         dns_response: &[u8],
         response_packet: &mut [u8],
     ) -> Result<usize, String> {
-        let length = ihl + UDP_HEADER + dns_response.len();
-        let buffer_slice = response_packet
-            .get_mut(ihl..length)
-            .ok_or_else(|| String::from("Out of bounds on response packet buffer"))?;
+        match self.payload {
+            PayloadRequestInfo::Udp {
+                source_port,
+                destination_port,
+                ..
+            } => {
+                let length = ihl + UDP_HEADER + dns_response.len();
+                let buffer_slice = response_packet
+                    .get_mut(ihl..length)
+                    .ok_or_else(|| String::from("Out of bounds on response packet buffer"))?;
+
+                let mut udp_response = MutableUdpPacket::new(buffer_slice).ok_or_else(|| {
+                    String::from("Failed to build MutableUdpPacket for response packet")
+                })?;
+
+                udp_response.set_source(destination_port);
+                udp_response.set_destination(source_port);
+                udp_response.set_length((UDP_HEADER + dns_response.len()) as u16);
+                udp_response.set_payload(dns_response);
+                udp_response.set_checksum(0);
+                udp_response
+                    .set_checksum(self.ip.compute_udp_checksum(&udp_response.to_immutable()));
 
-        let mut udp_response = MutableUdpPacket::new(buffer_slice)
-            .ok_or_else(|| String::from("Failed to build MutableUdpPacket for response packet"))?;
+                telio_log_debug!("UDP response: {:?}", &udp_response);
+                Ok(length)
+            }
+            PayloadRequestInfo::Tcp {
+                source_port,
+                destination_port,
+            } => {
+                let length = ihl + TCP_MIN_HEADER;
+                let buffer_slice = response_packet
+                    .get_mut(ihl..length)
+                    .ok_or_else(|| String::from("Out of bounds on response packet buffer"))?;
+                let mut tcp_response = MutableTcpPacket::new(buffer_slice).ok_or_else(|| {
+                    String::from("Failed to build MutableTcpPacket for response packet")
+                })?;
 
-        udp_response.set_source(self.udp.destination_port);
-        udp_response.set_destination(self.udp.source_port);
-        udp_response.set_length((UDP_HEADER + dns_response.len()) as u16);
-        udp_response.set_payload(dns_response);
-        udp_response.set_checksum(0);
-        udp_response.set_checksum(self.ip.compute_udp_checksum(&udp_response.to_immutable()));
+                tcp_response.set_source(source_port);
```

Comment:
I think we are missing sequence number for TCP

Reply:
fixed